ARROW-1463: [JAVA - WIP] refactoring

siddharthteotia opened the pull request:
cc @jacques-n , @BryanCutler , @icexelloss , @elahrvivaz

Prototype code (WIP) for BaseNullableFixedWidthVector abstract class and the implementation of subclass NullableIntVector.

ValueVector -> FixedWidthVector -> BaseNullableFixedVector (implements FieldVector) -> NullableIntVector

(1) No mutators and accessors
(2) Minimal delegation -- directly works with buffers for validity and actual data.
(3) No code generation
(4) Majority of the functionality abstracted out in the base class.
(5) Short call stack on hot paths.
(6) Buffer sharing is not yet done -- planning to first flush out the basic interfaces and get the infrastructure ready with new hierarchy.
(7) Need to figure out how to implement getFieldBuffers, loadFieldBuffers and other related methods that rely on vectors being BufferBacked.

siddharthteotia commented:
Pushed new changes based on comments on design spec.

elahrvivaz commented:
would it make sense to have some of the basic getter/setting methods pulled out to the base class with type parameters? or would that be bad for performance... i.e.
```
BaseNullableFixedWidthVector<T> {
  public void set(int index, T value);
  public T get(int index);
}
```
One of the pain points I've experienced is no consistent get/set methods, so when working with different vectors they all have to be cast to the exact type.

icexelloss commented:
@siddharthteotia This looks good so far. What I am concerned with are code duplication and maintainability if we hand write all the vector class. Can you maybe show (or describe) how `NullableLongVector` and `IntVector` would work? Specially for those setter/getter method.

I think @wesm suggested not having non-nullable/nullable vectors be two separate class to reduce the number of classes. We should discussion more on this.

icexelloss commented:
@elahrvivaz I think boxing/unboxing will be a performance issue if we use generics. So that's not an option here?

icexelloss commented:
@siddharthteotia 

nit: Can you reformat the description in markdown? You can do list like this:

* Something
* Something else

jacques-n commented:
@icexelloss: My preference is also to avoid having separate int and nullable int vectors. Only have NullableIntVector.  As far as duplication goes, I think the second commit here seems to be small enough number of things that we could skip the templates and just manage a little more code.

@elahrvivaz I'm not entirely clear on how the casting helps with generics. one of the holder methods could be made to follow this pattern. Can you provide an example pattern where the generic setter will help avoid any casting? (Or are you just talking about less casting?) We built the reader/writer interfaces specifically to avoid casting (in cases where people don't use generated code).

wesm commented:
should we create a wip-java-refactor where this work can take place without disrupting master? I presume we will require multiple patches

icexelloss commented:
@jacques-n I see. Yeah the `NullableIntVector` doesn't seem manageable.

jacques-n commented:
@icexelloss, I was saying the NullableIntVector does seem manageable. Were you agreeing or disagreeing?

icexelloss commented:
@jacques-n Ahh, sorry, there was a typo! Yes I agree.

elahrvivaz commented:
@jacques-n I haven't been able to use the writer framework, as it doesn't support all the operations we're doing (like dictionary encoding and fixed size lists, if I remember correctly). Our main data structure is a GeoTools SimpleFeature, which is basically just an `Array<Object>` with well-defined types at each index. For a lot of our work, I can set things up front to be the right types, and then say have an array of `Accessor`s and call `.getObject()` on them. However, `Mutator` doesn't have a generic setter method, so instead I have to cast down to the exact implementing class to call all the set methods.
@icexelloss possibly we could have the generic boxed methods, and specific unboxed methods in each impl? For me specifically I have to box everything anyway, but maybe that's not worth complicating the api for

BryanCutler commented:
Wow, this is great so far!  I really like that you were able to able to simplify rid of `Accessor` and `Mutator` and a +1 for me for having a single `NullableIntVector` class without codegen.  I think these are fairly stable now, so maintenance for all the vector types shouldn't be a big headache.

Is it possible we could rename vector classes too `NullableIntVector` to just `IntVector` since there is just one now, or is too confusing?

BryanCutler commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java:
```diff
@@ -0,0 +1,500 @@
+   protected int valueCount;
+
+   public BaseNullableFixedWidthVector (final String name, final BufferAllocator allocator,
+                                        FieldType fieldType, final int typeWidth) {
```
Is it possible to get `typewidth` from `fieldtype`?  Or maybe add a mappng in `ArrowType` that you could query here?

BryanCutler replied:
Nvm, I guess we need the TYPE_WIDTH in the concrete class for get/set methods anyway

elahrvivaz commented on java/vector/src/main/java/org/apache/arrow/vector/NullableIntVector.java:
```diff
@@ -0,0 +1,122 @@
+   }
+
+   @Override
+   public Types.MinorType getMinorType() {
```
this might be a good time to remove the MinorTypes and just use FieldTypes - I believe @jacques-n discussed this a while ago, that they don't actually capture all the field information.

BryanCutler replied:
If we could add a `getType` to return the `ArrowType` obj, that would be great.  Something like what #972 proposed to avoid having to cast would be helpful. But I'd hate to add more work to this, maybe better to discuss after the main stuff is done.

icexelloss replied:
I also found `MinorType` to be weird. But it's probably hard to remove due to compability? We could consider deprecating those though.

icexelloss replied:
Added to requirement 7 in the requirement doc. Let's keep this out of this PR and discuss separately?

siddharthteotia replied:
Yes, best to discuss separately. We will be merging multiple multiple patches to the new branch "java vector refactor" . We can do this once rest of the new code is stabilized, existing functionality is not broken etc. 

Later on we can merge this branch into master once refactoring is done or this particular thing can be addressed directly on master once refactor branch has been merged. I am open to ideas.

siddharthteotia commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java:
```diff
@@ -0,0 +1,500 @@
+      target.valueBuffer = valueBuffer.slice(startPoint, sliceLength).transferOwnership(target.allocator).buffer;
+   }
+
+   public void splitAndTransferValidityBuffer(int startIndex, int length,
```
Fixed.

siddharthteotia commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java:
```diff
@@ -0,0 +1,500 @@
+
+   private final String name;
+   private final BufferAllocator allocator;
+   private final int typeWidth;
```
Fixed.

siddharthteotia commented:
Notes on recent changes:

Addressed some comments. Some in progress.

- The current patch has the basic implementation of interfaces for Nullable Fixed and Variable Width vectors with corresponding concrete implementation of respective subclasses - NullableIntVector, NullableVarCharVector.

- Question: do we need allocationMonitor in our vector code? It complicates the logic of allocateNewSafe() and realloc().

- Question: do we need setting reader and writer index on arrow bufs inside vectors? I think it doesn't really matter and vector get/set methods should still work fine.

siddharthteotia commented:
My initial thoughts are to solidify this prototype -- flush out all APIs with complete implementation and tests and address comments. Right now we have concrete implementation for NullableInt and NullableVarChar in this prototype.

Can we have a separate project trunk/branch where we could merge patches for this work? Later on we can move the changes to Arrow master branch?

As we do this, we also need to think about getting rid of existing concrete implementation of vectors that is code generated. For example, once this prototype patch gets merged, we will need to remove NullableInt and NullableVarchar classes that are code generated.

cc @wesm, @jacques-n

icexelloss commented on java/vector/src/main/codegen/templates/NullableValueVectors.java:
```diff
@@ -105,6 +112,7 @@
     </#if>
     this.mutator = new Mutator();
     this.accessor = new Accessor();
+    <#if minor.class != "Int" && minor.class != "VarChar">
```
I wonder is it easier to have a separate template for Legency Vectors instead of if statements?

siddharthteotia replied:
I am not sure about that. I am actually thinking if instead of having legacy vectors, can we just put all the new vectors in a different package. Changing the template code to generate legacy vectors and do delegations to new vectors is like a throwaway work because eventually we will get rid of template all-together.

siddharthteotia replied:
nvm, I have made the changes to template to delegate the ops for Nullable Int and Nullable VarChar. Building  things now locally and trying to run tests.

icexelloss commented on java/vector/src/main/codegen/templates/NullableValueVectors.java:
```diff
@@ -44,6 +50,7 @@
  * NB: this class is automatically generated from ${.template_name} and ValueVectorTypes.tdd using FreeMarker.
  */
 @SuppressWarnings("unused")
+@Deprecated
```
Add "Use XXX instead" in deprecation warning?

siddharthteotia replied:
Sure.

icexelloss commented on java/vector/src/main/codegen/templates/UnionVector.java:
```diff
@@ -436,7 +436,11 @@ public Object getObject(int index) {
           <#if !minor.typeParams?? >
       case ${name?upper_case}:
-        return get${name}Vector().getAccessor().getObject(index);
+        <#if name != "INT" && name != "VARCHAR">
```
UnionVector is another mess. Should we keep UnionVector out of ARROW-1463?

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java:
```diff
@@ -0,0 +1,627 @@
+
+   /* used to step down the memory allocation */
+   protected void incrementAllocationMonitor() {
+      if (allocationMonitor < 0) {
```
Maybe refactor `allocationMonitor` to it's own class?

siddharthteotia replied:
I am trying to err on the side of less objects and less inner classes. So that is why thought we can maintain the allocationMonitor state through a simple counter. What do you think?

icexelloss replied:
I think it's fine to not have a object, but maybe put the allocationMonitor logic in base class?

siddharthteotia replied:
Right now we have other vector types inheriting from BaseValueVector. Once the entire hierarchy under BaseValueVector is refactored, we can move few things to BaseValueVector. There are a few things which are duplicated between BaseNullableFixedWidthVector and BaseNullableVariableWidthVector. Later on they can be moved to top level base class  -- BaseValueVector.

icexelloss replied:
In that case, I am leaning towards refactor allocationMonitor logic to a standalone class (in the logic is indeed the same among different vectors). I'd rather not maintaining the same logic in many different places...

siddharthteotia replied:
Let me clarify. I am aware that there is handful logic which is duplicated between two abstract classes only -- BaseNullableFixedWidthVector and BaseNullableVariableWidthVector. This is intentional for the time being.

It is not spread across several classes. We can move this shared logic to top level class BaseValueVector once the entire hierarchy has been refactored. 

For example, we haven't touched classes like ListVector and other which also inherit from BaseValueVector.

siddharthteotia replied:
I am just trying to phase out the work. Does the above explanation make some sense?

icexelloss replied:
Yes it does make sense. It's fine. We can always refactor in a later patch if we want to.

BryanCutler replied:
That sounds good to me @siddharthteotia to move it to a top level class later.  Do you think you could add a comment in the places that you anticipate changing in later phases?

siddharthteotia replied:
Sure, I will add note(s) in the code. Thanks @BryanCutler  and @icexelloss

siddharthteotia replied:
Done.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java:
```diff
@@ -0,0 +1,627 @@
+    *
+    * @return the number of null elements.
+    */
+   public int getNullCount() {
```
Is this function not performance critical so it's fine to recompute?

siddharthteotia replied:
Sorry, I don't understand. Are you saying we memoize the output of getNullCount()?

icexelloss replied:
Yeah, kind of, something like keep a `nullCount` variable in the class

icexelloss replied:
It might be surprising to user this function is not `O(1)`

siddharthteotia replied:
Ok, let's think about it.

icexelloss replied:
I don't think it's necessarily import for this PR to resolve this. Can you put a TODO for this?

siddharthteotia replied:
Sure.

siddharthteotia replied:
Done.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableVariableWidthVector.java:
```diff
@@ -0,0 +1,707 @@
+    *
+    * @return the number of null elements.
+    */
+   public int getNullCount() {
```
This function seems to be same as `getNullCount` in fix length vector

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableVariableWidthVector.java:
```diff
@@ -0,0 +1,707 @@
+    ******************************************************************/
+
+
+   protected void fillHoles(int index) {
```
Should these be `final prcotected`?

siddharthteotia replied:
Yes, that is correct. I missed it.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableVariableWidthVector.java:
```diff
@@ -0,0 +1,707 @@
+      }
+   }
+
+   private static byte getBitsFromCurrentByte(ArrowBuf data, int index, int offset) {
```
Is this common/well-defined enough to be moved to `BitVectorHelper`?

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/NullableIntVector.java:
```diff
@@ -0,0 +1,274 @@
+   }
+
+   @Override
+   protected org.slf4j.Logger getLogger() {
```
Hmm..Do we need the logger here?

icexelloss replied:
Nvm I see different vector has different logger.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/NullableIntVector.java:
```diff
@@ -0,0 +1,274 @@
+      valueBuffer.setInt(index * TYPE_WIDTH, value);
+   }
+
+   private void handleSafe(int index) {
```
This method seems generic enough, can this be in base class?

siddharthteotia replied:
Yes, you are right.

icexelloss commented on java/vector/src/test/java/org/apache/arrow/vector/TestBufferOwnershipTransfer.java:
```diff
@@ -60,8 +60,8 @@ public void testTransferVariableidth() {
 
     NullableVarCharVector v1 = new NullableVarCharVector("v1", childAllocator1);
     v1.allocateNew();
-    v1.getMutator().setSafe(4094, "hello world".getBytes(), 0, 11);
```
Is `getMutator` removed from the new API?

siddharthteotia replied:
Yes, one of the goals is to simplify the access and reduce the number of objects/classes. So all the getters and setters are directly accessible.

icexelloss replied:
Ok, the existing Spark code won' compile with this then. @BryanCutler or I can make the change in Spark, but in general I wonder if this is too disruptive.

icexelloss replied:
Doubling the number of getter/setter API is also not great, maybe we can add deprecated method and remove them in the next two releases or after 0.8 or sth. This is more a question for @wesm and @jacques-n to if we are willing to break user code in 0.8.

siddharthteotia replied:
I think the potential plan is to have the following:
Legacy Nullable <Type> Vector
Nullable <type> Vector -- the ones being implemented as part of refactoring

The latter one is not mutator/accessor based.

The existing consumers of these Nullable vectors should be able to continue using them through the LegacyNullable vector which is merely a delegation to new code for each API.

For example, if in Dremio code, we replace each reference to NullableIntVector with LegacyNullableIntVector, the code should still work without any surprises.

We can keep the legacy code around for now and then remove them all together (along with NullableValueVectors template) in the next release or so.

icexelloss replied:
After thinking about this a bit more, I feel maintaining backward compatibility of the new vectors might not to possible. The reasons are because I think we want to rename some vectors properly:

* DateDayVector, DateMillisVector ->DateVector
* IntervalDayVector -> IntervalVector
* Timestamp*Vector -> TimestampVector
* MapVector -> StructVector

Given this, we might as well just introduce new classes of vectors that is are slightly different (for instance, no accessor/mutator)

Also, I assume we would want ArrowReaders to return VectorSchemaRoot with new classes of vector in 0.8, is that correct?

Also,  ArrowWriters should work with both legency and new vectors?

icexelloss replied:
Btw, I updated the requirement doc and added the renaming requirement.

BryanCutler replied:
+1 on renaming. Are we also planning on removing the "Nullable" prefix to vectors?  like `NullableIntVector` -> `IntVector`.  I know that's currently what the inner vectors are called but after the refactor it won't really make sense to call it `NullableIntVector` because the inner vectors will be gone and all vectors are nullable right?

siddharthteotia replied:
Yes, that's probably right. @jacques-n , what do you think?

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BaseNullableFixedWidthVector.java:
```diff
@@ -0,0 +1,627 @@
+   protected abstract org.slf4j.Logger getLogger();
+
+   @Override
+   public Mutator getMutator() {
```
If we are going to remove Mutator/Accessor, we should remove these function so it's compiler error rather than runtime error for the user.

siddharthteotia replied:
Yes, for that we need to remove the generic Accessor, Mutator interface all together from ValueVector and its subinterfaces. 

Since Legacy vectors are having mutator/accessor based access -- we may still need these temporary 2 functions getMutator() and getAccessor() just to comply with the interface until Legacy vectors have been removed.

icexelloss commented on java/vector/src/main/java/org/apache/arrow/vector/BitVectorHelper.java:
```diff
@@ -0,0 +1,42 @@
+
+import io.netty.buffer.ArrowBuf;
+
+public class BitVectorHelper {
```
This class no public static methods?

icexelloss replied:
I found a public util class that has no public methods to be weird. Maybe making this class package private if that's the intention?

siddharthteotia replied:
Sounds good.

siddharthteotia replied:
Done.

siddharthteotia commented:
The patch finally builds fine :) Running existing tests now.

icexelloss commented:
@siddharthteotia Can you create a branch and change this PR to be merged to that branch?

wesm commented:
I created a `java-vector-refactor` branch for you to collaborate in https://git-wip-us.apache.org/repos/asf?p=arrow.git;a=shortlog;h=refs/heads/java-vector-refactor

siddharthteotia commented:
Closing this one.
